### PR TITLE
fix(builtins.provide): should only replace exactly matched expression

### PR DESCRIPTION
**Pull request by suxin2017**
### <samp>🤖 Generated by Copilot at 6e0af05</samp>

This pull request adds the `provide` feature to the `@rspack/core` package, which allows users to inject require calls for modules specified in the `provide` option of the `builtins` plugin. It implements the feature in the `rspack_plugin_javascript` crate, and adds tests and changesets for it. It also updates the `.vscode/settings.json` file to support the new crate.

<details open=true>
  <summary><h2>Walkthrough</h2></summary>

### <samp>🤖 Generated by Copilot at 6e0af05</samp>

*  Add a changeset file to document the patch update for the `@rspack/core` package ([link](https://github.com/web-infra-dev/rspack/pull/2721/files?diff=unified&w=0#diff-b78ade17c6bdca6277fbda61a40b3e0690b3282a360e30552b7b54a6a1e561f5R1-R5))
*  Update the `.vscode/settings.json` file to include the `./crates/rspack_plugin_javascript/Cargo.toml` file in the `rust-analyzer.linkedProjects` array ([link](https://github.com/web-infra-dev/rspack/pull/2721/files?diff=unified&w=0#diff-a5de3e5871ffcc383a2294845bd3df25d3eeff6c29ad46e3a396577c413bf357L33-R34))
  *  Add new imports for `std::collections::HashSet` and some AST types from `swc_core::ecma::ast` ([link](https://github.com/web-infra-dev/rspack/pull/2721/files?diff=unified&w=0#diff-ceac836b6796451c62fe37267c43d2739a75e6b9356c825ddce76c5900c2a18cL1-R8))
  *  Add new static constants `SOURCE_DOT` and `MODULE_DOT` and a new field `current_import_provide` to the `ProvideBuiltin` struct ([link](https://github.com/web-infra-dev/rspack/pull/2721/files?diff=unified&w=0#diff-ceac836b6796451c62fe37267c43d2739a75e6b9356c825ddce76c5900c2a18cL12-R20))
  *  Modify the `handle_ident` and `handle_member_expr` methods to check and add the provided identifiers to the `current_import_provide` set and create new identifiers with the replaced dot character and the unresolved mark ([link](https://github.com/web-infra-dev/rspack/pull/2721/files?diff=unified&w=0#diff-ceac836b6796451c62fe37267c43d2739a75e6b9356c825ddce76c5900c2a18cL23-R48))
  *  Add a new method `create_provide_require` to create a vector of `ModuleItem`s that represent the require calls for the provided modules and handle nested paths ([link](https://github.com/web-infra-dev/rspack/pull/2721/files?diff=unified&w=0#diff-ceac836b6796451c62fe37267c43d2739a75e6b9356c825ddce76c5900c2a18cL116-R219))
  *  Modify the `visit_mut_expr` and `visit_mut_module` methods to use the `create_provide_require` method and insert the require calls at the beginning of the module body ([link](https://github.com/web-infra-dev/rspack/pull/2721/files?diff=unified&w=0#diff-ceac836b6796451c62fe37267c43d2739a75e6b9356c825ddce76c5900c2a18cL116-R219))
  *  Add a new `visit_mut_var_decl` method to prevent visiting the variable declarations created by the `create_provide_require` method ([link](https://github.com/web-infra-dev/rspack/pull/2721/files?diff=unified&w=0#diff-ceac836b6796451c62fe37267c43d2739a75e6b9356c825ddce76c5900c2a18cL116-R219))
  *  Add the input code for testing the `provide` feature in the `index.js` and `process.js` files ([link](https://github.com/web-infra-dev/rspack/pull/2721/files?diff=unified&w=0#diff-c20b2d0e81223d90dcbb99e404820ec61e04b4d420afac879abac24a5f120e25L1-R1), [link](https://github.com/web-infra-dev/rspack/pull/2721/files?diff=unified&w=0#diff-5f8c20431c40675f2c6a70b5fe62323d187ff73c00bbc53327428a75bb14fa1eR1-R191))
  *  Add the expected output of the plugin in the `expected/main.js` file ([link](https://github.com/web-infra-dev/rspack/pull/2721/files?diff=unified&w=0#diff-35bca6b107cd0329c7a1af193d64eb596cb0afb6530cd1b866d8a3962047648fL1-R139))
  *  Add the configuration for testing the `provide` feature in the `test.config.json` file ([link](https://github.com/web-infra-dev/rspack/pull/2721/files?diff=unified&w=0#diff-56fc1acd6dd1a2bc2c34d996c2f05a3b6474a76587f80f1076611877f952851fL1-R9))
*  Add a new test case to the `packages/rspack/tests/configCases/builtins/provide/index.js` file that checks the variable coverage of the `provide` feature ([link](https://github.com/web-infra-dev/rspack/pull/2721/files?diff=unified&w=0#diff-e5a24d044c31e364d739eeaa5774a5b6c9aa45a30b526423c8e98e19f99734ffR52-R58))
*  Add the same code as the `process.js` file in the test fixtures to the `packages/rspack/tests/configCases/builtins/provide/process.js` file ([link](https://github.com/web-infra-dev/rspack/pull/2721/files?diff=unified&w=0#diff-849a43118b9f4eef14d8e948b5c9ddf89cb2fb0b5357213bfa945b253f69808eR1-R191))
*  Update the `packages/rspack/tests/configCases/builtins/provide/webpack.config.js` file to add the `process` identifier and the `./process.js` module path to the `provide` object of the `builtins` plugin ([link](https://github.com/web-infra-dev/rspack/pull/2721/files?diff=unified&w=0#diff-9626a8d60282348c76be8b5734fe62869c193a89addede40fec2cf9b342075f3L18-R19))

</details>

## Related issue (if exists)
https://github.com/web-infra-dev/rspack/issues/2590

## Types of changes

- [ ] Docs change / Dependency upgrade
- [x] Bug fix
- [ ] New feature / Improvement
- [ ] Refactoring
- [ ] Breaking change

## Checklist

- [x] I have added changeset via `pnpm run changeset`.
- [x] I have added tests to cover my changes.

### CODE REVIEW

**Comment by suxin2017** [comment]
```
"./index.js": function (module, exports, __webpack_require__) {
var bbb1 = __webpack_require__("./aaa.js")["c"];
var aaa1 = __webpack_require__("./aaa.js");
console.log(aaa);
var dddd = __webpack_require__("./aaa.js");
var dddd = 123;
console.log(aaa.bbb.ccc);
console.log(bbb);
},
```
The ` var bbb1 = __webpack_require__ (". / aaa. Js ") (" c "); `
It should be 'bbb' Where is the 'one' added? Can someone help me
@ahabhgk

**Comment by suxin2017** [comment]
> ```
> "./index.js": function (module, exports, __webpack_require__) {
> var bbb1 = __webpack_require__("./aaa.js")["c"];
> var aaa1 = __webpack_require__("./aaa.js");
> console.log(aaa);
> var dddd = __webpack_require__("./aaa.js");
> var dddd = 123;
> console.log(aaa.bbb.ccc);
> console.log(bbb);
> },
> ```
> 
> The `var bbb1 = __webpack_require__ (". / aaa. Js ") (" c ");` It should be 'bbb' Where is the '1' added? Can someone help me @ahabhgk

I see why this is. Can you ignore it

File: .vscode/settings.json [hunk 1]
```diff
@@ -30,7 +30,8 @@
   "rust-analyzer.procMacro.attributes.enable": true,
   "rust-analyzer.linkedProjects": [
     "Cargo.toml",
-    "crates/node_binding/Cargo.toml"
+    "crates/node_binding/Cargo.toml",
+    "./crates/rspack_plugin_javascript/Cargo.toml"
```
[the remark this answers is not in the document]
Comment (suxin2017) [reply .vscode/settings.json:34]: It's not expected. I'll fix it

**Review by hyf0 — APPROVED** [review]
Nice job. Thank you!

**Pull request merged**